Cross-file `declare module` interface augmentation not merged into keyof/indexed-access/assignability (fp-ts HKT `URItoKind` → false TS2344, extends #13509)

Goal: green

fp-ts is blocked by **854 false `TS2344`** ("`Type 'typeof URI' does not satisfy the constraint 'keyof URItoKind<any>'`") plus ~473 downstream `TS2322`/`TS2741`, all stemming from one root cause: when an interface is augmented cross-file via `declare module './X' { interface I<A> { ... } }` from a sibling file, tsz does **not** fold those members into the interface's materialized structural type, so `keyof URItoKind<any>` resolves to `never` and every HKT tag fails its constraint. This reproduces on the current binary (`/Users/mohsen/code/tsz-origin-main-check/.target/dist-fast/tsz`); `tsc` 6.0.3 accepts the same code (this is the canonical fp-ts higher-kinded-type / defunctionalization pattern). #13509 fixed the **same-module** augmentation path (`apply_self_module_augmentations`), but the **cross-file** members still never reach the solver's keyof/Application/assignability evaluation, so fp-ts emits the identical `TS2344` after #13509 was closed.

## Structural rule

When an interface receives a `declare module './X'` augmentation from a **different** file, `tsc` merges those members into the interface so that `keyof`, indexed access (`URItoKind<A>[U]`), and structural assignability all see them; tsz only merges **same-module** augmentations (the #13509 path), so cross-file `keyof URItoKind<any>` evaluates to `never`.

`When an exported interface `I` declared in home file `H` is augmented from a different file `F` via `declare module './H' { interface I<A> { ... } }`, tsc merges `F`'s members into `I`'s instantiated structural type at every materialization site (keyof, indexed access, assignability); tsz does X only for same-module augmentations, leaving the cross-file member set empty.`

Owner: **tsz-checker** symbol-type resolution, which feeds the (under-populated) member map to the solver's keyof / Application / structural-assignability queries.

Source sites:
- `crates/tsz-checker/src/state/type_resolution/symbol_types.rs:207` — the symbol-type resolution path calls `apply_self_module_augmentations(sym_id, structural_type)` (Step 1.26, "fp-ts HKT `URItoKind` pattern, #13509"). The merged result is not what the cross-file `keyof`/Application/assignability query reads.
- `crates/tsz-checker/src/types/module_augmentation.rs:1348` — `apply_self_module_augmentations`; it resolves the home file name and delegates to `apply_module_augmentations` (`:1164`). The relative-specifier key matching for the augmenting file (`'./HKT'`) lives in `module_augmentation_key_candidates` (`:18`), whose cross-file candidate set must include the augmenting-file → home-file mapping for the merge to fire.

## Minimal repro

Two files. `HKT.ts` declares the central interface and the `keyof`-derived HKT machinery; `Array.ts` augments it cross-file.

```ts
// HKT.ts
export interface URItoKind<A> {}
export type URIS = keyof URItoKind<any>;
export type Kind<U extends URIS, A> = U extends URIS ? URItoKind<A>[U] : never;
```

```ts
// Array.ts
import { Kind, URIS, URItoKind } from './HKT';

declare module './HKT' {
  interface URItoKind<A> {
    readonly MyArray: Array<A>;
  }
}

// 'MyArray' is a member of URItoKind cross-file, so keyof URItoKind<any> ⊇ 'MyArray'.
export type AllURIS = keyof URItoKind<any>;
export type MyArr = Kind<'MyArray', number>; // <- tsz: TS2344
```

Command:

```sh
/Users/mohsen/code/tsz-origin-main-check/.target/dist-fast/tsz \
  --noEmit --strict --target es2022 --lib es2022,dom \
  /tmp/fpts_hkt/HKT.ts /tmp/fpts_hkt/Array.ts
```

Actual tsz output (exit 2):

```
Array.ts(17,26): error TS2344: Type '"MyArray"' does not satisfy the constraint 'keyof URItoKind<any>'.
```

`tsc` 6.0.3 on the same two files: **exit 0** (accepts).

Faithful fp-ts form — `typeof URI` tag + computed `[URI]` key — produces the symbolic-`keyof` display variant (the 62 fp-ts errors that print `keyof URItoKind<any>` rather than the already-resolved `never`):

```ts
// Faithful.ts (with the same HKT.ts above)
import { URIS, Kind, URItoKind } from './HKT';
export const URI = 'MyArray';
export type URI = typeof URI;
declare module './HKT' {
  interface URItoKind<A> { readonly [URI]: Array<A>; }
}
export interface MyFunctor<F extends URIS> {
  readonly map: <A, B>(fa: Kind<F, A>, f: (a: A) => B) => Kind<F, B>;
}
export type Demo = MyFunctor<URI>; // tsz: TS2344 'typeof URI'; tsc: OK
```

```
Faithful.ts(18,30): error TS2344: Type 'typeof URI' does not satisfy the constraint 'keyof URItoKind<any>'.
```

Same-file control that **passes on both** (proves the bug is specifically cross-file `declare module`, not interface merging in general): plain same-file interface declaration merging, no `declare module`:

```ts
// SameFileMerge.ts
interface URItoKind<A> {}
interface URItoKind<A> { readonly MyArray: Array<A>; }
type URIS = keyof URItoKind<any>;
type Kind<U extends URIS, A> = U extends URIS ? URItoKind<A>[U] : never;
export const _x: Kind<'MyArray', number> = [1, 2, 3]; // tsz: OK, tsc: OK (both exit 0)
```

(Note: a same-file `declare module './SameFile' { interface URItoKind<A> {...} }` form is rejected by **both** tsc and tsz — that is parity, not a control, because tsc also does not self-augment a module from inside itself. The valid control is plain declaration merging above.)

## Relationship to #13509

#13509 was closed by PR #13515, which added `apply_self_module_augmentations` to "fold self-module augmentations into exported interfaces before keyof." That fix covers the **same-module** case: an augmentation whose home file is the file currently being resolved. It does **not** make cross-file augmentation members reach the solver's keyof / Application / assignability evaluation of the augmented interface — so the fp-ts pattern (each combinator module augments `./HKT` from its own sibling file) is still broken, and fp-ts emits the **identical** `TS2344` after #13509 closed. This issue extends #13509 to the cross-file path: same owner function family (`apply_self_module_augmentations` → `apply_module_augmentations` → `module_augmentation_key_candidates`), uncovered for the home-file→augmenting-file cross-file mapping.

## Affected corpus rows

- **fp-ts** (`gcanti/fp-ts` @ `c0a64721`, `src`): **dominant first blocker.** 854 false `TS2344` (792 print `constraint 'never'`, 62 print the symbolic `keyof URItoKind<any>` — same root cause, only the type-printer display form differs depending on whether the `keyof` was already resolved to `never`) plus ~473 downstream `TS2322`/`TS2741` cascading from the empty `URItoKind`. Real source: `HKT.ts:49` `interface URItoKind<A> {}`, `HKT.ts:80` `type URIS = keyof URItoKind<any>`, `ReadonlyArray.ts:1855` `declare module './HKT'`.
- **io-ts** (`Kind2` / `URItoKind2`, the 2-ary member of the same HKT family) will likely benefit from this fix once its separate blocker (a missing fp-ts dependency in the fixture `node_modules`, surfacing as `TS2307`) is also resolved. io-ts is **not** included as an affected row here because its first blocker is the missing dep, not this keyof gap.

## Adjacent cases to cover

- Cross-file 1-ary augmentation (`URItoKind<A>`) **and** 2-ary (`URItoKind2<E, A>` / `Kind2<F, E, A>`).
- Literal-key augmentation (`readonly MyArray: ...`) **and** computed-`[URI]`-key augmentation (`readonly [URI]: ...`).
- Renamed URI symbols (the binder name of the tag const must not drive the fix — vary it).
- Augmentation declared in a **third** file (home `HKT.ts`, augmenter `Array.ts`, consumer `use.ts`).
- Member consumed via **keyof** (`keyof URItoKind<any>`), via **indexed access** (`URItoKind<A>[U]`), and via **structural assignability**: `const r: URItoKind<number> = {}` must be **TS2741** (missing `MyArray`) — tsz currently **accepts** this (exit 0) while tsc emits `TS2741: Property 'MyArray' is missing in type '{}' but required in type 'URItoKind<number>'`, directly demonstrating the augmented members are absent from the materialized structural type cross-file.

## Suggested fix (performance-aware)

Merge cross-file augmentation members into the target interface's **materialized structural type at every use site** (keyof, Application/indexed-access, assignability) — not only the import-alias path. Currently `apply_self_module_augmentations` (`module_augmentation.rs:1348`) handles same-module augmentation but the cross-file home-file→augmenting-file mapping does not feed the merged member set into the solver queries.

Performance plan (a fix that regresses throughput on augmentation-free programs is unacceptable):
1. Keep the `program_has_module_augmentations()` fast-path early-out (it is already the first guard in `apply_self_module_augmentations`) so non-augmenting programs pay nothing.
2. Memoize the merged interface type per `(DefId, type-args)` so augmentation collection + `merge_properties` runs **once per instantiation**, not per keyof/index/relation query.
3. Reuse the existing `module_augmentation_application_set` re-entrancy guard (`module_augmentation.rs:1186`/`:1323`).
4. The cross-file lookup must match the home-file → augmenting-file relative-specifier (`'./HKT'`) candidate via `module_augmentation_key_candidates` (`module_augmentation.rs:18`) — i.e. resolve each augmentation key from its augmenting file back to the home file's source index and merge when they coincide.

Anti-hardcoding: the fix must be purely structural (DefId / module-graph / member-map driven). Do not key on `URItoKind`, `URI`, file names, or rendered type text.

## Verification

```sh
TSZ_PROJECT_COMPILE_SET=canary TSZ_PROJECT_COMPILE_FILTER='^fp-ts-project$' \
  bash scripts/ci/project-compile-guard.sh
```

Expect the fp-ts `TS2344` count to drop to **0** (and the downstream `TS2322`/`TS2741` to clear). Add a targeted `tsz-checker` unit test exercising cross-file `URItoKind` augmentation through `keyof` (and the 2-ary `URItoKind2`/`Kind2` form), with the binder name of the augmenting tag varied to guard against name-keyed regressions, plus the `TS2741` assignability assertion (`const r: URItoKind<number> = {}`).

---
*Surfaced by the grow-corpus re-survey (2026-06-14) against a fresh `dist-fast` build of `main`; root-caused and adversarially verified via multi-agent analysis (repro run against the binary, tsc behavior cross-checked). Tracked under corpus-readiness #13512.*

<!-- Provenance: Mohsens-MacBook-Pro / claude-code / claude-opus-4-8 / ultracode -->

--- a/crates/tsz-checker/Cargo.toml
+++ b/crates/tsz-checker/Cargo.toml
@@ -521,6 +521,9 @@ path = "tests/export_star_module_augmentation_tests.rs"
 name = "hkt_self_augmentation_keyof_repro"
 path = "tests/hkt_self_augmentation_keyof_repro.rs"
 [[test]]
+name = "hkt_cross_file_augmentation_13653_repro"
+path = "tests/hkt_cross_file_augmentation_13653_repro.rs"
+[[test]]
 name = "cjs_object_export_module_augmentation_merge_tests"
 path = "tests/cjs_object_export_module_augmentation_merge_tests.rs"
 [[test]]

--- a/crates/tsz-checker/src/state/type_analysis/core.rs
+++ b/crates/tsz-checker/src/state/type_analysis/core.rs
@@ -1315,6 +1315,52 @@ impl<'a> CheckerState<'a> {
         let (result, type_params) = self.compute_type_of_symbol(sym_id);
         self.pop_symbol_dependency();
 
+        // Fold cross-file `declare module` augmentations into an exported
+        // interface's materialized body at this canonical resolution point, so
+        // every downstream cache (symbol_types, both type environments, the
+        // shared def store) observes the SAME augmented body. Doing it only on
+        // the type-reference path (`type_reference_symbol_type`) left this path —
+        // reached for the solver's def-store registration and cross-file
+        // delegation — publishing the un-augmented body, which then shadowed the
+        // augmented form at every `keyof` / Application / assignability site
+        // (#13653, extends the same-module #13509 fix to the cross-file path).
+        // `apply_self_module_augmentations` is a no-op unless the program has
+        // augmentations and the symbol is an exported, non-imported interface.
+        // The `program_has_module_augmentations` guard keeps augmentation-free
+        // programs from paying for the symbol lookup on every interface.
+        //
+        // Only merge when `result` is ALREADY a concrete object/callable shape:
+        // a `Lazy`/unresolved cross-file body (e.g. when this symbol's home file
+        // is not the file currently being checked) would otherwise fall into the
+        // augmentation intersection fallback and cache a degenerate type. Those
+        // cases are handled by the import/type-reference path once the body
+        // resolves; here we only need to keep the *materialized* body augmented
+        // so it cannot shadow that path's result in the shared def store.
+        // Cheap checks first: the program-wide augmentation guard, then the
+        // interface-flag test, and only then the per-`TypeId`
+        // `classify_for_augmentation` lookup — so non-interface symbols never pay
+        // for the classification.
+        let result = if self.ctx.program_has_module_augmentations()
+            && self
+                .ctx
+                .binder
+                .get_symbol(sym_id)
+                .or_else(|| self.get_cross_file_symbol(sym_id))
+                .is_some_and(|symbol| {
+                    symbol.has_any_flags(symbol_flags::INTERFACE)
+                        && !symbol.has_any_flags(symbol_flags::CLASS)
+                })
+            && matches!(
+                crate::query_boundaries::common::classify_for_augmentation(self.ctx.types, result),
+                crate::query_boundaries::common::AugmentationTargetKind::Object(_)
+                    | crate::query_boundaries::common::AugmentationTargetKind::ObjectWithIndex(_)
+                    | crate::query_boundaries::common::AugmentationTargetKind::Callable(_)
+            ) {
+            self.apply_self_module_augmentations(sym_id, result)
+        } else {
+            result
+        };
+
         // Pop from resolution stack
         if use_local_symbol_state {
             self.ctx.symbol_resolution_stack.pop();

--- a/crates/tsz-checker/src/types/module_augmentation.rs
+++ b/crates/tsz-checker/src/types/module_augmentation.rs
@@ -719,8 +719,8 @@ impl<'a> CheckerState<'a> {
                     if (member_node.kind == PROPERTY_SIGNATURE
                         || member_node.kind == METHOD_SIGNATURE)
                         && let Some(sig) = arena.get_signature(member_node)
-                        && let Some(name_node) = arena.get(sig.name)
-                        && let Some(id_data) = arena.get_identifier(name_node)
+                        && let Some(member_name) =
+                            self.augmentation_member_key_name(arena, sig.name)
                     {
                         let type_id = if std::ptr::eq(arena, self.ctx.arena) {
                             let mut type_id = if member_node.kind == PROPERTY_SIGNATURE
@@ -752,7 +752,7 @@ impl<'a> CheckerState<'a> {
 
                         aug_member_order += 1;
                         members.push(PropertyInfo {
-                            name: self.ctx.types.intern_string(&id_data.escaped_text),
+                            name: self.ctx.types.intern_string(&member_name),
                             type_id,
                             write_type: type_id,
                             optional: sig.question_token,
@@ -1327,6 +1327,66 @@ impl<'a> CheckerState<'a> {
         result
     }
 
+    /// Resolve an augmentation member's property-key name from the augmenting
+    /// file's own arena/binder, purely syntactically.
+    ///
+    /// Plain identifier, string-literal, and numeric-literal keys are read
+    /// directly. A computed key `[expr]` is resolved when `expr` is a string
+    /// literal or an identifier bound to a string-initialized `const` in the
+    /// augmenting file — the fp-ts `const URI = "X"; interface I { [URI]: ... }`
+    /// pattern. The resolution is deliberately syntactic (no expression type
+    /// evaluation): augmentation member collection runs *inside* interface body
+    /// resolution, so calling the evaluating `get_property_name_resolved` here
+    /// would re-enter interface resolution and recurse. It also works uniformly
+    /// for cross-arena augmentations whose key `const` lives in another file's
+    /// binder (#13653).
+    fn augmentation_member_key_name(
+        &mut self,
+        arena: &tsz_parser::parser::NodeArena,
+        name_idx: tsz_parser::parser::NodeIndex,
+    ) -> Option<String> {
+        Self::augmentation_member_key_name_in_arena(arena, name_idx, |ident_name| {
+            let binder = self.ctx.get_binder_for_arena(arena)?;
+            let sym_id = binder.file_locals.get(ident_name)?;
+            binder
+                .get_symbol(sym_id)
+                .and_then(|symbol| arena.get(symbol.value_declaration))
+                .and_then(|decl| arena.get_variable_declaration(decl))
+                .and_then(|var_decl| arena.get(var_decl.initializer))
+                .and_then(|init| arena.get_literal(init))
+                .map(|lit| lit.text.clone())
+        })
+    }
+
+    /// Pure-arena key resolution shared by all augmentation paths.
+    /// `resolve_const` maps a computed-key identifier to its string-`const`
+    /// value (so the same matching can be unit-tested without a live binder).
+    fn augmentation_member_key_name_in_arena(
+        arena: &tsz_parser::parser::NodeArena,
+        name_idx: tsz_parser::parser::NodeIndex,
+        resolve_const: impl FnOnce(&str) -> Option<String>,
+    ) -> Option<String> {
+        use tsz_parser::parser::syntax_kind_ext::COMPUTED_PROPERTY_NAME;
+        let name_node = arena.get(name_idx)?;
+        if let Some(ident) = arena.get_identifier(name_node) {
+            return Some(ident.escaped_text.clone());
+        }
+        if let Some(lit) = arena.get_literal(name_node) {
+            return Some(lit.text.clone());
+        }
+        if name_node.kind == COMPUTED_PROPERTY_NAME {
+            let computed = arena.get_computed_property(name_node)?;
+            let expr_node = arena.get(computed.expression)?;
+            if let Some(lit) = arena.get_literal(expr_node) {
+                return Some(lit.text.clone());
+            }
+            if let Some(ident) = arena.get_identifier(expr_node) {
+                return resolve_const(ident.escaped_text.as_str());
+            }
+        }
+        None
+    }
+
     /// Merge `declare module "./home" { interface I { ... } }` augmentations into
     /// the type of interface `I` declared (and exported) in its own home module,
     /// even when `I` is reached by reference within its declaring file rather than
@@ -1472,6 +1532,70 @@ mod tests {
     use std::sync::Arc;
     use tsz_binder::{BinderState, ModuleAugmentation};
     use tsz_parser::parser::{NodeArena, ParserState};
+
+    /// `augmentation_member_key_name_in_arena` resolves identifier, string-literal,
+    /// and computed string-`const` property keys, and leaves an unresolvable
+    /// computed key as `None`. The const resolver is renamed (`TAG`, not a
+    /// hard-coded fp-ts name) to lock the structural, name-independent contract.
+    #[test]
+    fn augmentation_member_key_name_resolves_identifier_string_and_computed_const() {
+        let source = r#"
+interface I {
+    foo: number;
+    "bar": number;
+    [TAG]: number;
+    [OTHER]: number;
+}
+"#;
+        let mut parser = ParserState::new("t.ts".to_string(), source.to_string());
+        parser.parse_source_file();
+        let arena = Arc::new(parser.into_arena());
+        let sf = arena.source_files.first().expect("source file");
+        let iface_idx = sf
+            .statements
+            .nodes
+            .iter()
+            .copied()
+            .find(|&idx| {
+                arena
+                    .get(idx)
+                    .and_then(|n| arena.get_interface(n))
+                    .is_some()
+            })
+            .expect("interface node");
+        let iface = arena
+            .get_interface(arena.get(iface_idx).expect("iface node"))
+            .expect("interface data");
+
+        let resolved: Vec<Option<String>> = iface
+            .members
+            .nodes
+            .iter()
+            .copied()
+            .filter_map(|member_idx| {
+                let member = arena.get(member_idx)?;
+                let sig = arena.get_signature(member)?;
+                Some(CheckerState::augmentation_member_key_name_in_arena(
+                    &arena,
+                    sig.name,
+                    // Only `TAG` is a known string const; `OTHER` is unknown.
+                    |name| (name == "TAG").then(|| "computed_tag".to_string()),
+                ))
+            })
+            .collect();
+
+        assert_eq!(
+            resolved,
+            vec![
+                Some("foo".to_string()),
+                Some("bar".to_string()),
+                Some("computed_tag".to_string()),
+                None,
+            ],
+            "key resolver must handle identifier, string-literal, and computed \
+             string-const keys, and drop unresolvable computed keys"
+        );
+    }
 
     #[test]
     fn module_augmentation_has_type_params_detects_type_alias_with_params() {

--- a/crates/tsz-checker/tests/hkt_cross_file_augmentation_13653_repro.rs
+++ b/crates/tsz-checker/tests/hkt_cross_file_augmentation_13653_repro.rs
@@ -1,0 +1,219 @@
+//! Repro + adjacent matrix for #13653: a cross-file `declare module` interface
+//! augmentation must be folded into the interface's materialized structural type
+//! at every use site — `keyof`, indexed access, AND structural assignability —
+//! not only the same-module path covered by #13509.
+//!
+//! Structural rule: when an exported interface `I` declared in home file `H` is
+//! augmented from a different file via `declare module "./H" { interface I { ... } }`,
+//! `keyof I`, `I[K]`, and assignability against `I` must all observe the merged
+//! members. The fix folds augmentations into the interface body at the canonical
+//! `get_type_of_symbol` resolution point so every shared def-store / type-env
+//! cache observes the same augmented body across files.
+//!
+//! These use `check_all_multi_file_with_global_index`, which checks every file
+//! through one shared definition store — the multi-file path where an
+//! un-augmented body registered while checking a consumer would otherwise shadow
+//! the augmented body (the regression this guards).
+
+use tsz_checker::context::CheckerOptions;
+use tsz_checker::test_utils::check_all_multi_file_with_global_index;
+
+fn diagnostics(files: &[(&str, &str)]) -> Vec<(u32, String)> {
+    check_all_multi_file_with_global_index(
+        files,
+        CheckerOptions {
+            strict: true,
+            ..Default::default()
+        },
+    )
+    .into_iter()
+    .map(|d| (d.code, d.message_text))
+    .collect()
+}
+
+fn count_code(diags: &[(u32, String)], expected: u32) -> usize {
+    diags.iter().filter(|(code, _)| *code == expected).count()
+}
+
+/// Core fp-ts HKT witness: a 1-ary registry augmented cross-file, consumed via a
+/// conditional `Kind` that indexes the registry.
+#[test]
+fn cross_file_augmentation_seen_by_keyof_and_indexed_access() {
+    let diags = diagnostics(&[
+        (
+            "HKT.ts",
+            r#"
+export interface URItoKind<A> {}
+export type URIS = keyof URItoKind<unknown>;
+export type Kind<U extends URIS, A> = U extends URIS ? URItoKind<A>[U] : never;
+"#,
+        ),
+        (
+            "Array.ts",
+            r#"
+import { Kind, URItoKind } from "./HKT";
+declare module "./HKT" {
+    interface URItoKind<A> {
+        readonly MyArray: ReadonlyArray<A>;
+    }
+}
+export type MyArr = Kind<"MyArray", number>;
+"#,
+        ),
+    ]);
+
+    assert_eq!(
+        count_code(&diags, 2344),
+        0,
+        "cross-file augmented key missed by keyof/indexed-access; got {diags:#?}"
+    );
+}
+
+// Note: the computed-`[URI]`-key augmentation form (the faithful fp-ts `typeof
+// URI` tag) is exercised end-to-end via the CLI / fp-ts project-compile-guard
+// (see PR Verification) and unit-tested at the name-resolution layer in
+// `crates/tsz-checker/src/types/module_augmentation.rs`
+// (`augmentation_member_key_name_in_arena`). It is omitted from this lib-less
+// multi-file harness, which cannot materialize the lib-backed computed-key
+// member shape the real driver resolves.
+
+/// Anti-hardcoding: rename every binder (module, registry interface, key, alias);
+/// the rule is structural, not name-driven.
+#[test]
+fn cross_file_augmentation_is_binder_name_independent() {
+    let diags = diagnostics(&[
+        (
+            "registry.ts",
+            r#"
+export interface Slots<T> {}
+export type Tags = keyof Slots<unknown>;
+export type Lookup<K extends Tags, T> = K extends Tags ? Slots<T>[K] : never;
+"#,
+        ),
+        (
+            "widget.ts",
+            r#"
+import { Lookup, Slots } from "./registry";
+declare module "./registry" {
+    interface Slots<T> {
+        readonly Widget: ReadonlyArray<T>;
+    }
+}
+export type W = Lookup<"Widget", number>;
+"#,
+        ),
+    ]);
+
+    assert_eq!(
+        count_code(&diags, 2344),
+        0,
+        "renamed-binder registry should also see the augmented key; got {diags:#?}"
+    );
+}
+
+/// Augmentation declared in a THIRD file: home declares, a sibling augments, a
+/// consumer (neither home nor augmenter) reads the augmented key.
+#[test]
+fn cross_file_augmentation_visible_from_third_consumer_file() {
+    let diags = diagnostics(&[
+        (
+            "HKT.ts",
+            r#"
+export interface URItoKind<A> {}
+export type URIS = keyof URItoKind<unknown>;
+export type Kind<U extends URIS, A> = U extends URIS ? URItoKind<A>[U] : never;
+"#,
+        ),
+        (
+            "Array.ts",
+            r#"
+import { URItoKind } from "./HKT";
+declare module "./HKT" {
+    interface URItoKind<A> {
+        readonly MyArray: ReadonlyArray<A>;
+    }
+}
+export {};
+"#,
+        ),
+        (
+            "use.ts",
+            r#"
+import { Kind } from "./HKT";
+import "./Array";
+export type MyArr = Kind<"MyArray", number>;
+"#,
+        ),
+    ]);
+
+    assert_eq!(
+        count_code(&diags, 2344),
+        0,
+        "third-file consumer should see the augmentation; got {diags:#?}"
+    );
+}
+
+/// Structural assignability: the cross-file augmented member is REQUIRED, so an
+/// empty object literal assigned to the interface is TS2741 (missing member).
+/// tsz previously accepted this (the augmented member was absent from the
+/// materialized structural type cross-file).
+#[test]
+fn cross_file_augmentation_member_is_required_for_assignability() {
+    let diags = diagnostics(&[
+        (
+            "HKT.ts",
+            r#"
+export interface URItoKind<A> {}
+"#,
+        ),
+        (
+            "Array.ts",
+            r#"
+import { URItoKind } from "./HKT";
+declare module "./HKT" {
+    interface URItoKind<A> {
+        readonly MyArray: ReadonlyArray<A>;
+    }
+}
+export const r: URItoKind<number> = {};
+"#,
+        ),
+    ]);
+
+    assert_eq!(
+        count_code(&diags, 2741),
+        1,
+        "cross-file augmented member should be required (TS2741); got {diags:#?}"
+    );
+}
+
+/// Negative control (#6164): a NON-exported file-local interface that is
+/// self-augmented via `declare module "./self"` must NOT absorb the augmented
+/// member, so an object literal carrying that member stays an excess property
+/// (TS2353). Guards against over-merging.
+#[test]
+fn non_exported_self_augmented_interface_does_not_merge() {
+    let diags = diagnostics(&[(
+        "test.ts",
+        r#"
+interface Local {
+    a: number;
+}
+
+declare module "./test" {
+    interface Local {
+        b: string;
+    }
+}
+
+const v: Local = { a: 1, b: "x" };
+export {};
+"#,
+    )]);
+
+    assert_eq!(
+        count_code(&diags, 2353),
+        1,
+        "non-exported local interface must keep its self-augmentation isolated; got {diags:#?}"
+    );
+}

